### PR TITLE
Fallback on a browser if no search app is installed

### LINKED ISSUE #158: Search online without Google App
Would it be possible to have the "Search online" feature of the search bar available without having the Google App installed?
My suggestion would be to launch the Google App if installed and fall back on just launching a Google search in the Browser if not.

### PR DESCRIPTION
**Pull request by flocke**
This is my implementation of a fallback online search using the ACTION_VIEW intent in case no search app is installed.
The URI used for this search can be configured in the settings, by default Google is used.

Fixes  #158

### CODE REVIEW

File: app/src/main/java/com/benny/openlauncher/activity/Home.java [hunk 1]
```diff
@@ -501,8 +501,16 @@ private void initSearchBar() {
...
-                Intent intent = new Intent(Intent.ACTION_WEB_SEARCH);
-                intent.putExtra(SearchManager.QUERY, string);
+                Intent intent = null;
+
+                if (Tool.isIntentAvailable(getApplicationContext(), Intent.ACTION_WEB_SEARCH)) {
+                    intent = new Intent(Intent.ACTION_WEB_SEARCH);
+                    intent.putExtra(SearchManager.QUERY, string);
+                } else {
+                    intent = new Intent(Intent.ACTION_VIEW);
+                    intent.setData(Uri.parse(appSettings.getSearchBarBaseURI()+string));
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/Home.java]: Please don't just append, this implies that search query must be at the end, which is not always the case.
Add something like an tag `{%s}`, if that is really not available you can fallback to string append.

File: app/src/main/java/com/benny/openlauncher/activity/Home.java [hunk 2]
```diff
@@ -501,8 +501,16 @@ private void initSearchBar() {
         searchBar.setCallback(new SearchBar.CallBack() {
             @Override
             public void onInternetSearch(String string) {
-                Intent intent = new Intent(Intent.ACTION_WEB_SEARCH);
-                intent.putExtra(SearchManager.QUERY, string);
+                Intent intent = null;
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/Home.java]: You can initialize with `new Intent()` already here, and use `setAction` then

File: app/src/main/java/com/benny/openlauncher/util/Tool.java [hunk 3]
```diff
@@ -615,4 +616,15 @@ public static void copy(Context context, String stringIn, String stringOut) {
             Toast.makeText(context, R.string.dialog__backup_app_settings__error, Toast.LENGTH_SHORT).show();
         }
     }
+
+    public static boolean isIntentAvailable(Context context, String action) {
+        final PackageManager packageManager = context.getPackageManager();
+        final Intent intent = new Intent(action);
+        List resolveInfo = packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
+        if (resolveInfo.size() > 0) {
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/util/Tool.java]: `return resolveInfo.size() > 0;`

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -91,6 +92,9 @@
     <string name="pref_summary__desktop_show_position_indicator">Show the page indicator below the desktop.</string>
     <string name="pref_title__desktop_show_label">Show Labels</string>
     <string name="pref_summary__desktop_show_label">Show item labels below their icons.</string>
+    <string name="pref_title__search_bar_base_uri">Search URI</string>
+    <string name="pref_summary__search_bar_base_uri">Base URI for web searches (the search term will be appended at the end).</string>
+    <string name="pref_default__search_bar_base_uri" translatable="false">https://google.com/search?q=</string>
```
Comment (gsantner) [inline app/src/main/res/values/strings.xml]: See above

File: app/src/main/res/xml/preferences_desktop.xml [hunk 5]
```diff
@@ -59,6 +59,17 @@
...
     </PreferenceCategory>
 
+    <PreferenceCategory android:title="@string/pref_title__search_bar">
+
+        <EditTextPreference
+            android:defaultValue="@string/pref_default__search_bar_base_uri"
+            android:key="@string/pref_key__search_bar_base_uri"
+            android:summary="@string/pref_summary__search_bar_base_uri"
+            android:title="@string/pref_title__search_bar_base_uri"
+            android:inputType="textUri" />
```
Comment (gsantner) [inline app/src/main/res/xml/preferences_desktop.xml]: You could also set the default value additionally as `android:hint` so it gets visible if the input gets cleared

**Comment by flocke** [comment]
Thanks for your replies, I will implement all the stuff you pointed out.
Since I'm very new to Java, what would you think is the best way to do the string replacement for the search query?

**Comment by gsantner** [comment]
`url = base.contains("{%s}") ? base.replace("{%s}", query) : (base + query);`
It can also be done with some more if/else, but thats the easiest way

**Comment by flocke** [comment]
I've fixed almost everything.
Do you know how I can make it so that an empty entry for the search URI in the settings falls back to the default?

**Comment by gsantner** [comment]
That will automatically happen if user presses the Reset App button

**Comment by flocke** [comment]
Yes, but right now with the hint set it shows the default URI when you delete all text in the settings item.
I fear that could confuse some users to think the default one will be used, when in fact just an empty string is stored (and used).

**Comment by gsantner** [comment]
Yes, but grayed out. In the case where an bad URL is stored the intent will throw an exception anyway, in this case you could use the default value again.

**Comment by gsantner** [comment]
Thanks for your work! Congrats on your first merge on OL :tada:

**Comment by gsantner** [comment]
Thanks for your work! 
Congrats on your first merged code contributions to OL :tada:

**Comment by flocke** [comment]
Thank you for the quick response and all the comments :thumbsup:

**Comment by mase76** [comment]
I have configured to use duckduckgo with
https://duckduckgo.com/?q={query}
However google is still used in the browser.

**Comment by flocke** [comment]
@mase76 That sounds like your browser supports the WEB_SEARCH intent. In that case the URL is not used and you have to change the default search engine in your browser.

_Sent from my Google Nexus 5X using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by mase76** [comment]
No, ddg is my default search engine in firefox. I have no search app installed.

**Comment by gsantner** [comment]
@mase76 
You have to distinguish between the WebSearch and the "address bar"-search. The address bar uses your setted search engine, while WebSearch is fixed set to Google at Firefox Android.

**Comment by gsantner** [comment]
The implementation of WEB_SEARCH is up to the app consuming it. Also, Intent docs explictly mentions Google for WEB_SEARCH intent. See <https://developer.android.com/reference/android/content/Intent.html#ACTION_WEB_SEARCH>. I don't know which app you got installed, but there is some which consumes the intent. gplay services / Google Now for example.

**Comment by mase76** [comment]
I have omnirom without gapps. No search app installed. Firefox is my browser.

**Comment by flocke** [comment]
Firefox consumes the WEB_SEARCH intent as well.
And it seems it uses Google for that and not the search engine you configured.

**Comment by flocke** [comment]
@gsantner I think we should have an option to always use the VIEW intent with the configured URL even if a search app is installed. What do you think? I can get a pull request ready if you want.

**Comment by mase76** [comment]
Then my issue can't be fixed. Or openlauncher must call the browser with a url, as configured in ol.

**Comment by gsantner** [comment]
>  I can get a pull request ready if you want.

It's an option. But I would like to suggest to wait because theres still refactor work going on.

> Then my issue can't be fixed. Or openlauncher must call the browser with a url, as configured in ol.

For clarification: The url search stuff is intended as **fallback** for the regular serach intent.

**Comment by mase76** [comment]
I posted too slow. Yes, that would be an option.  expected that behaviour from the beginning, as I configured the url with the query.

**Comment by flocke** [comment]
@gsantner Ok, I'm gonna wait until the refactor is done.
@mase76 For now you could install the DuckDuckGo app and use that for the WEB_SEARCH intent (until my pull request is ready).

**Comment by mase76** [comment]
Ok, thx!

**Comment by flocke** [comment]
@gsantner How far along is the refactor?

**Comment by gsantner** [comment]
>  How far along is the refactor?

@MFlisar

**Comment by MFlisar** [comment]
@gsantner 
For now it's finished (and merged already). Until now I've not encountered any problem.

**Comment by flocke** [comment]
@mase76 My pull request should be included in the newest release. You now have an option to always use the browser (and with it the configured URI) for searches. Just enable it in the OL settings.

_Sent from my Google Nexus 5X using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Pull request merged**